### ISSUE TITLE
restrictions on Google digitized content prevent adding new works using API 

### ISSUE THREAD

**Issue by mnaydan** [Princeton-CDH/ppa-django#339]
## testing notes
The interface for adding new items is the same as before: log in to django admin, select Digitized Works, click '+ Add from HathiTrust'. Test various real/possible scenarios, including:
- add a single item
- add multiple items
- try to add something with an invalid/mistyped id
- try to add something that's already in ppa
- add a public domain google-digitized book

For things that are successfully added, check that they have been added and indexed correctly. Look at the admin view, public view, and try finding them in the public search. Searches on book metadata and page text should both work;  you can check that pages are indexed by using the 'search within a volume' feature. There may be a small lag between when you add something and when it's searchable, but if it isn't searchable after several minutes, something probably went wrong.  On the admin view, check the 'history' page to confirm there's a log message documenting when the record was created.

I expect there will be practical limitations with how many items you can add at a time, depending on the size of the items, based on how long rsync and indexing takes, but I don't know how many it will be. You could experiment to try to get a sense, and if necessary we can tweak the timeout configuration on the server. 

## dev notes
Remove code for working with HathiTrust Data API and use dataset rsync instead 

- [x] add method to `HathiImporter` to rsync a list of ids into the local hathi data pairtree files; see https://github.com/hathitrust/datasets/wiki/Dataset-rsync-instructions#sync-a-list-of-ids
- [x] revise `DigitizedWork.add_from_hathi` method to remove option & logic for getting data
- [x] remove `DigitizedWork.get_hathi_data` method entirely
- [x] remove `HathiDataAPI` entirely from `ppa.archive.hathi`
- [x] revise `HathiImporter.add_items` method to run new rsync method before adding to the database; drop argument to get data when adding

**Comment by rlskoeser** [comment]
The relevant view for this work is [AddFromHathiView](https://github.com/Princeton-CDH/ppa-django/blob/da14a7ca85585948ac3e37991c07dc16bec77818/ppa/archive/views.py#L616), which hands off to a [HathiImporter](https://github.com/Princeton-CDH/ppa-django/blob/master/ppa/archive/util.py) utility class. Possibly we only need to modify the `add_items` method on the importer.

Items are currently added individually using the [static method add_from_hathi](https://github.com/Princeton-CDH/ppa-django/blob/da14a7ca85585948ac3e37991c07dc16bec77818/ppa/archive/models.py#L686) on DigitizedWork. For rsync, it probably makes more sense to add them in bulk instead of individually like we needed to do with the data api.  The `add_from_hathi` method should be revised to remove the option to download data, and I think the `get_hathi_data` method  should go away entirely. We should make sure the new implementation preserves the logic to add django admin log messages (should be handled by existing `add_from_hathi` logic).

We'll need a new method — maybe on the importer utility class — to rsync data for a list of ids. We'll have to adapt the examples from the rsync dataset instructions here: https://github.com/hathitrust/datasets/wiki/Dataset-rsync-instructions#sync-a-list-of-ids (would be nice to contribute python instructions back if we have or can make it reusable).

After data is synced, the logic will be: use `add_from_hathi` to add to the database, and then index the pages in Solr.  (Equivalent to running `hathi_import` and `index` manage commands.)

Doing all this via http is going to be slow for large number of documents, but we're not going to worry about that for now; assume small numbers of documents. (But we should think about error handling in case of interruptions or a large document.)

**Comment by kmcelwee** [comment]
Testing rsync on QA caused error. We don't have permission on that server to sync with hathi.

On production, the following commands were run:
```bash
# make and move to temporary directories
cd /home/deploy/repos/Princeton-CDH/ppa-django
mkdir -p tmp/test_data
cd tmp

# prepare variables
export TREE=ht_text_pd
export LOCAL_PATH=test_data

# create file with one id path
echo "hvd/pairtree_root/hn/2w/xb/hn2wxb" > path_list.txt

# from hathi's instructions:
rsync --copy-links --delete --ignore-errors --recursive --times --verbose --files-from=path_list.txt datasets.hathitrust.org::$TREE $LOCAL_PATH
```

returned the following error:
```
rsync: failed to connect to datasets.hathitrust.org (134.68.125.241): Connection timed out (110)
rsync: failed to connect to datasets.hathitrust.org (134.68.125.243): Connection timed out (110)
rsync error: error in socket IO (code 10) at clientserver.c(125) [Receiver=3.1.2]
```

**Comment by kmcelwee** [comment]
Connection timed out for `nyp/pairtree_root/33/43/30/82/51/56/48/33433082515648` as well as a garbage path: `foo/ba/12`, meaning we can't rule out that our paths are malformed.

**Comment by kmcelwee** [comment]
The rsync flag `--ignore-existing : skip updating files that already exist on receiver` may be useful

**Comment by mnaydan** [comment]
tl;dr: 1) non-Google digitized works aren't being indexed properly (main keyword search works, but search within volume not working); note this is NOT happening in production, but IS happening to other non-Google digitized works (not just recently added) on test site. 2) adding limited-search only HT ID breaks site (500 error).

- add single item (non-Google) works (uc2.ark:/13960/t3416vd1p). Creates record in backend, appears in backend log, and is findable on public site through keyword, author, title searches. BUT, search within volume isn't working ("edible frog"). A keyword search yields a certain number of matching hits on the main page, but when I click "see more details" the search yields no results.

- add single item (Google-digitized) works (pst.000003850619). Weirdly, search within volume for this one ("nayle") DOES work! I tried going back to first record to see if it was a lag, but it's still not working. I think I've figured out it has to do with non-Google digitized books: (same thing happening with coo1.ark:/13960/t0pr8bh4d (recently added) and dul1.ark:/13960/t5j970j9d (already in PPA)).

- trying to add an invalid ID yields "Failed to sync data" on backend.

- trying to add item already in PPA (hvd.32044102846144) yields "Skipped; already in database"

- One case I thought of that wasn't in testing notes was if we try to add a Hathi ID that is limited-search only (njp.32101073249110): so the ID is valid, in Hathi, but not in public domain or full text searchable. This yields a 500 server error.  Seems like it should be on us to make sure we're adding things only in public domain, but also, if this happens, it probably shouldn't break the site? After breaking, it does create a full metadata record in the backend but doesn't appear to be findable on public site.

- Add multiple records at once works. Added four valid public domain records - took about 10 seconds. I tried adding 20 records, a mix of valid, invalid, and already in PPA (including suppressed). Took around 35 seconds. Skipped, failed, and added records as expected.

**Comment by rlskoeser** [comment]
@mnaydan thanks for the detailed testing feedback.  

The search within volume error for uc2.ark:/13960/t3416vd1p was actually a syntax error in the search filter rather than an indexing problem; I think it was because of the format of the volume identifier. I've made a change to address it.

Thanks for testing the limited-search only id; the test I had for rsync failure didn't work in this case, because it syncs the metadata but not the page data! Added another check, which should report this as an rsync failure.

When you get the chance, please test again and see if everything is working now.

**Comment by mnaydan** [comment]
@rlskoeser it's failing to sync data now for full text public domain Google and non-Google digitized books that we should be able to add. For instance: mdp.39015082250914 or uiuo.ark:/13960/t4fn7694k.

**Comment by rlskoeser** [comment]
whoops! my check for the zip file must be wrong

**Comment by rlskoeser** [comment]
@mnaydan I adjusted the path I'm checking to see if the zip file was synced correctly. Would you test again?

**Comment by mnaydan** [comment]
@rlskoeser huh, so I added a list of 9 items, a mix of all the different types. It loaded for a while and then sent me to a 500 server error page. I clicked back, tried to submit the list again, and then it said all the valid IDs were in the database already. And all the valid IDs were indeed successfully added in the previous attempt it seems, despite generating the 500 error. I'll have to test again tomorrow when I have more time to see if this was just a fluke or what happened.

**Comment by rlskoeser** [comment]
@mnaydan I think it choked on this id: "bc.ark:/13960/t7xm3bv6p" — the code is generating a "not a valid pairtree" error, which is a new one to me! I have a thought about what might be causing it... Is this the first bc item you're adding? 

(My suspicion is that it's creating a brand new pairtree section in our local storage, and it didn't initialize properly.)

**Comment by mnaydan** [comment]
@rlskoeser yes, so some of the non-google digitized books have unusual IDs (usually ark:/...) and in searching for those, I found a few that have various letters before the ark:/... (like uiuo.ark:/ above, or this bc.ark:/ one)

**Comment by rlskoeser** [comment]
@mnaydan ok! I've made another adjustment, to include a few more files in cases where (like this one) you are adding an item with a new prefix that we didn't have before. If you could, please test with an id that has a new prefix (the letters before the `.`) that we don't already have in the site to confirm I'm handling that case correctly now.

More details, if you're interested: the letters before the `.` are a library/institution prefix (I think it's the MARC institution code, e.g. Princeton is njp), and each institution gets own pairtree folder within the HathiTrust data. To be a valid pairtree, there are two files required at the top level — a prefix and a version. I wasn't including those in the list of files to be synced, so adding content that wasn't in one of our existing pairtrees resulted in the 'not a valid pairtree' error.

**Comment by mnaydan** [comment]
Everything looks good! Re-tested add one, add multiple, non-Google valid ID, Google valid ID, limited search only, already in PPA, invalid ID, and new prefix (ku01.r2_308), and all outcomes behaved as desired. I double checked that search by metadata and keyword, including search inside volume, works, and that the admin records were created. 

Fascinating about the pairtree - I'm so glad we caught that! Thank you for your iterative work on this feature!!
